File: ch3/clip5/boj1261/Main.java
import java.util.PriorityQueue;
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        final int INF = 1000000000;
        int m = sc.nextInt(), n = sc.nextInt();
        int[][] graph = new int[n + 1][m + 1];
        int[][] cost = new int[n + 1][m + 1];
        for(int i = 1; i <= n; i++) {
            char[] input = sc.next().toCharArray();
            for(int j = 1; j <= m; j++) {
                graph[i][j] = input[j - 1] - '0';
                cost[i][j] = INF;
            }
        }
        PriorityQueue<Point> pq = new PriorityQueue<>((o1, o2) -> {
            return o1.cost - o2.cost;
        });
        int[] dr = {-1, 1, 0, 0};
        int[] dc = {0, 0, -1, 1};
        pq.offer(new Point(1, 1, 0));
        cost[1][1] = 0;
        while(!pq.isEmpty()) {
            Point now = pq.poll();
            if(cost[now.row][now.col] < now.cost) continue;
            if(now.row == n && now.col == m) break;
            for(int i = 0; i < 4; i++) {
                int nr = now.row + dr[i];
                int nc = now.col + dc[i];
                if(nr < 1 || nr > n || nc < 1 || nc > m) continue;
                if(cost[nr][nc] > cost[now.row][now.col] + graph[nr][nc]) {
                    cost[nr][nc] = cost[now.row][now.col] + graph[nr][nc];
                    pq.offer(new Point(nr, nc, cost[nr][nc]));
                }
            }
        }
        System.out.println(cost[n][m]);
    }
}

class Point {
    int row, col, cost;
    public Point(int row, int col, int cost) {
        this.row = row;
        this.col = col;
        this.cost = cost;
    }
}

File: ch3/clip5/boj1261/main.py
import sys
import heapq

INF = 1000000000

m, n = map(int, sys.stdin.readline().split())
graph = []
cost = [[INF] * m for _ in range(n)]
cost[0][0] = 0

for _ in range(n):
    row = list(map(int, list(sys.stdin.readline().strip())))
    graph.append(row)

dr = [-1, 1, 0, 0]
dc = [0, 0, -1, 1]

pq = [(0, 0, 0)]  # 튜플로 요소를 추가

while pq:
    now = heapq.heappop(pq)
    row, col, current_cost = now  # 요소를 추출

    if current_cost > cost[row][col]:
        continue

    if row == n - 1 and col == m - 1:
        break

    for i in range(4):
        nr = row + dr[i]
        nc = col + dc[i]

        if nr < 0 or nr >= n or nc < 0 or nc >= m:
            continue

        new_cost = current_cost + graph[nr][nc]

        if new_cost < cost[nr][nc]:
            cost[nr][nc] = new_cost
            heapq.heappush(pq, (nr, nc, new_cost))  # 튜플로 요소를 추가

sys.stdout.write(str(cost[n - 1][m - 1]))
